4.0.0rc1 Problem with CCDData.__init__

Currently the `CCDData.__init__` checks if the wcs is in fact a `WCS`. This leads to a failure in `ccdproc` because after slicing the wcs it's a SlicedWCS which makes creating the sliced `CCDData` fail:

```
====================================================== FAILURES =======================================================
____________________________________________ test_trim_with_wcs_alters_wcs ____________________________________________

    def test_trim_with_wcs_alters_wcs():
        ccd_data = ccd_data_func()
        # WCS construction example pulled form astropy.wcs docs
        wcs = WCS(naxis=2)
        wcs.wcs.crpix = np.array(ccd_data.shape)/2
        wcs.wcs.cdelt = np.array([-0.066667, 0.066667])
        wcs.wcs.crval = [0, -90]
        wcs.wcs.ctype = ["RA---AIR", "DEC--AIR"]
        wcs.wcs.set_pv([(2, 1, 45.0)])
        ccd_wcs = CCDData(ccd_data, wcs=wcs)
        # The trim below should subtract 10 from the 2nd element of crpix.
        # (Second element because the FITS convention for index ordering is
        #  opposite that of python)
>       trimmed = trim_image(ccd_wcs[10:, :])

ccdproc\tests\test_ccdproc.py:313:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
...\lib\site-packages\astropy\nddata\mixins\ndslicing.py:63: in __getitem__
    return self.__class__(**kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = CCDData([[-0.74882747,  0.56759473,  0.71815054, ...,  1.30935097,
           0.69587966,  0.32030821],
         [-0.6... -0.82190772],
         [-0.66845311, -2.15776154, -0.88186588, ..., -0.47226641,
           0.58196437,  0.97061286]])
args = ()
kwd = {'data': array([[-0.74882747,  0.56759473,  0.71815054, ...,  1.30935097,
         0.69587966,  0.32030821],
       [-...      0.58196437,  0.97061286]]), 'mask': None, 'meta': {'my_key': 42, 'your_key': 'not 42'}, 'uncertainty': None, ...}
llwcs = SlicedLowLevelWCS Transformation

This transformation has 2 pixel and 2 world dimensions

Array shape (Numpy order): N...elation between pixel and world axes:

           Pixel Dim
World Dim    0    1
        0  yes  yes
        1  yes  yes

    def __init__(self, *args, **kwd):
        if 'meta' not in kwd:
            kwd['meta'] = kwd.pop('header', None)
        if 'header' in kwd:
            raise ValueError("can't have both header and meta.")

        super().__init__(*args, **kwd)
        if self._wcs is not None:
            llwcs = self._wcs.low_level_wcs
            print(type(llwcs), llwcs)
            if not isinstance(llwcs, WCS):
>               raise TypeError("the wcs must be a WCS instance.")
E               TypeError: the wcs must be a WCS instance.

...\lib\site-packages\astropy\nddata\ccddata.py:195: TypeError
```

That's a bug that I probably introduced when we made `NDData` require a APE-14 WCS.

I'm not sure if this is a release-blocker, but it definitely needs fixing.

--- a/astropy/nddata/ccddata.py
+++ b/astropy/nddata/ccddata.py
@@ -199,6 +199,19 @@ class CCDData(NDDataArray):
         if _config_ccd_requires_unit and self.unit is None:
             raise ValueError("a unit for CCDData must be specified.")
 
+    def _slice_wcs(self, item):
+        """
+        Override the WCS slicing behaviour so that the wcs attribute continues
+        to be an `astropy.wcs.WCS`.
+        """
+        if self.wcs is None:
+            return None
+
+        try:
+            return self.wcs[item]
+        except Exception as err:
+            self._handle_wcs_slicing_error(err, item)
+
     @property
     def data(self):
         return self._data

--- a/astropy/nddata/tests/test_ccddata.py
+++ b/astropy/nddata/tests/test_ccddata.py
@@ -1021,3 +1021,13 @@ def test_read_returns_image(tmpdir):
     ccd = CCDData.read(filename, unit='adu')
     # Expecting to get (5, 5), the size of the image
     assert ccd.data.shape == (5, 5)
+
+
+# https://github.com/astropy/astropy/issues/9664
+def test_sliced_ccdata_to_hdu():
+    wcs = WCS(naxis=2)
+    wcs.crpix = 10, 10
+    ccd = CCDData(np.ones((10, 10)), wcs=wcs, unit='pixel')
+    trimmed = ccd[2:-2, 2:-2]
+    hdu = trimmed.to_hdu()
+    assert isinstance(hdu, fits.HDUList)

--- a/astropy/nddata/tests/test_ccddata.py
+++ b/astropy/nddata/tests/test_ccddata.py
@@ -1026,8 +1026,10 @@ def test_read_returns_image(tmpdir):
 # https://github.com/astropy/astropy/issues/9664
 def test_sliced_ccdata_to_hdu():
     wcs = WCS(naxis=2)
-    wcs.crpix = 10, 10
+    wcs.wcs.crpix = 10, 10
     ccd = CCDData(np.ones((10, 10)), wcs=wcs, unit='pixel')
     trimmed = ccd[2:-2, 2:-2]
-    hdu = trimmed.to_hdu()
-    assert isinstance(hdu, fits.HDUList)
+    hdul = trimmed.to_hdu()
+    assert isinstance(hdul, fits.HDUList)
+    assert hdul[0].header['CRPIX1'] == 8
+    assert hdul[0].header['CRPIX2'] == 8

